### ISSUE TITLE
Performance and Laziness

### ISSUE THREAD

**Issue by ccorcos** [hyperapp/hyperapp#75]
It appears there's no laziness at all in hyperapp -- no diff of props:

https://github.com/hyperapp/hyperapp/blob/f29647bf9da4bc281078abcacd116031f9fb0c3e/src/app.js#L159-L163

Have you tried this out in a large application with 1000s of nodes?

**Comment by jbucaran** [comment]
There's also no batching for vnodes at the moment, but there [used to be](https://github.com/hyperapp/hyperapp/blob/b5cc01f06b5e62d1ad1a78b7d9b073369ce5dd05/index.js#L28). There are more than a couple things we could do to optimize the vdom implementation.

Here's a [repaint perf test](http://mathieuancelin.github.io/js-repaint-perfs/hyperapp/).

> there's no laziness at all
Would that be something you can help with?

See also:
* https://github.com/hyperapp/hyperapp/issues/13

**Comment by leeoniya** [comment]
@jbucaran please consider implementing https://github.com/krausest/js-framework-benchmark, it is much more representative of real-world performance.

the only thing dbmonster tests is initial creation and `.className` and `.textContent` (or `.nodeValue`) replacement; it doesnt test complex reconciliation cases, track-by-key bindings or dom mutations that cause reflow/restyle.

**Comment by jbucaran** [comment]
@leeoniya Thanks! That's next on my list.

**Comment by ccorcos** [comment]
> Can you help with that?

Sure! Let me see what I can do.

**Comment by ccorcos** [comment]
oh wait, I just realized there's no composition of components here...

**Comment by jbucaran** [comment]
@ccorcos What do you mean?

**Comment by tunnckoCore** [comment]
@jbucaran maybe #2, haha. @ccorcos ?

**Comment by ccorcos** [comment]
yes, every similar. there needs to be some way of combining components together. Hyper isnt so scalable in terms of breaking up components and reusing them by combining them together. For example:

```js
const TwoCounters = app({
    model: {
        one: Counter.model,
        two: Counter.model,
    },
    update: {
        one: ({one, two}) => ({two, one: Counter.update(one)})
        two: ({one, two}) => ({one, two: Counter.update(two)})
    },
    view: ({one, two}, actions) =>
        <div>
            {Counter.view(one, actions.one)}
            {Counter.view(one, actions.two)}
        </div>
})
```

Then laziness could be introduced where Counter one doesnt need to be rerendered when Counter two changes...

**Comment by jbucaran** [comment]
@ccorcos This is not supported since https://github.com/hyperapp/hyperapp/pull/77.

```
<MyComponent propts=...>children</MyComponent>
```

**Comment by jbucaran** [comment]
I got news. I'm using https://wildlyinaccurate.com/todomvc-perf-comparison as the benchmark runner. I'm not yet done cleaning up the implementation, but  here are some preliminary results:

<img width="1000" alt="screen shot 2017-02-17 at 23 57 05" src="https://cloud.githubusercontent.com/assets/56996/23070070/30280002-f56d-11e6-9a57-fc56b8400a3e.png">

Clearly there's much work to be done to improve the vdom engine. Currently no optimizations are in place, so I expect these numbers to improve over time as I work on the engine, but at least we are not the slowest.

**Comment by tunnckoCore** [comment]
We are almost like Angular, it's not okey. But yea.

Can you add `choo`? So we can see how we are compared to it which is similar and uses real dom.

**Comment by leeoniya** [comment]
from that screenshot, there's not a single recent version lib there; the fastest impl (Mithril) is using an ancient 0.1.21.

take my word for it, just bench this: https://github.com/krausest/js-framework-benchmark

**Comment by jbucaran** [comment]
@tunnckoCore This is time consuming, so I can't promise I'll add it soon, I'

@leeoniya I just forked lhorie's implementation of the benchmark and added mine, I wasn't really paying attention to the version of the other libraries.

Do you have any idea how to get started with the krausest/js-framework-benchmark benchmarks? The README does an incredible good job at not explaining that and from all the other benchmarks I've seen, this one looks more challenging.

**Comment by leeoniya** [comment]
> Do you have any idea how to get started with the krausest/js-framework-benchmark benchmarks?

Yes. The implementation itself is easy, though actually running the bench is less easy - i usually leave the latter to @krausest since he does this for the majority of pull requests and merges.

There are two versions of the bench, and you can implement one or both: `-keyed` and `-non-keyed`. Here are examples of my impls with domvm [1][2]. Your `package.json` should have a `build-prod` script [3] that bundles and compiles your app as needed.

Other than that, all you have to make sure is that it behaves correctly in the browser, like `vanillajs` and `vanillajs-keyed`. Youc an get an idea of the perf from the console timers that are called from your impl, but the actual bench will extract the numbers from Chrome's timeline in a very stable/reproducible way between runs.

[1] https://github.com/krausest/js-framework-benchmark/tree/master/domvm-v2.0.1-keyed/src
[2] https://github.com/krausest/js-framework-benchmark/tree/master/domvm-v2.0.1-non-keyed/src
[3] https://github.com/krausest/js-framework-benchmark/blob/master/domvm-v2.0.1-keyed/package.json#L7

**Comment by jbucaran** [comment]
Also, I'm not sure what it means, but FWIW HyperApp was actually the _fastest_ in the async version of the benchmark. 😄

**Comment by leeoniya** [comment]
> You clearly are aware of something I am not.

I've done impls of a lot of benchmarks. This one is by far the most representative of expected perf and highly stable between runs due to its methodology. Basically, this bench tends to accurately reflect the performance results of all other benches and covers enough surface area not to only test a couple specific cases (like dbmon). There's nothing TodoMVC bench will tell you that you won't find out more thoroughly and with better feedback from `js-framework-benchmark`.

There are other good ones, like uibench and vdom-benchmark but they don't have the most up-to-date impls of other libs, so it's difficult to compare vs most recent lib versions. They also dont have great output for assessing relative perf in an intuitively-colored scale.

> Also, I'm not sure what it means, but FWIW HyperApp was actually the fastest in the async version of the benchmark. 😄

If your implementation or lib is async, rather than relying on the console timers, you should open up chrome's timeline view and record the perf between actions. Then look at "Scripting time", etc in the pie chart.

**Comment by krausest** [comment]
I added some hints how to start your implementation in the README of js-framework-benchmark at the bottom. Just let me know if that helps. If you have specific questions don't hesitate to ask.

**Comment by ccorcos** [comment]
@jbucaran I bet the bigger the app, the worse hyper will perform. try that benchmark with a website that has 1000s of components

**Comment by tunnckoCore** [comment]
@ccorcos of course, absolutely, because each "app" can sucks more than the framework with which it is built. :D

**Comment by ccorcos** [comment]
well my point is if you wanted to build a custom visualization or some kind of dashboard, the lack of laziness is going to become a problem. i dont mean to be a put-down -- I like what you guys are building here

**Comment by jbucaran** [comment]
@ccorcos Do laziness only applies to components? I revisited your example and now I get what you were saying. 👍

**Comment by dodekeract** [comment]
Related question: does hyperapp currently re-use old DOM nodes when rendering?

**Comment by jbucaran** [comment]
/cc @ccorcos @tunnckoCore @dodekeract @sotojuan

**tl;dr**

* HyperApp has **decent** performance.
* Optimized engines like Inferno and Mithril are unequivocally faster, but there is _no_ extraordinary advantage. 
* Choo has consistently the poorest performance of all the tested frameworks.

---

I followed @leeoniya's suggestion and implemented the js-framework-benchmarks. I still need to create the PR, but I wanted to share the results right away.

HyperApp's numbers in the TodoMVC benchmarks before were alright-ish, but some runs actually yielded better numbers than others. 

Why is that? I suspect my TodoMVC implementation is half the issue. `Object.assign` is heavily used to clone list items and there are virtually no mutations. The other half, I presume, is deleting items, like the graph below shows.

### Results
![](https://cloud.githubusercontent.com/assets/56996/23337555/5c4eda10-fc35-11e6-8459-7f37bea648e5.png)
![](https://cloud.githubusercontent.com/assets/56996/23337556/5e4cf428-fc35-11e6-8250-9ee7b7a3f6e3.png)

### Hardware

![](https://cloud.githubusercontent.com/assets/56996/23337559/a598d180-fc35-11e6-9afc-bdec1fb31be0.png)


### Notes

* Compare with @krausest's [results here](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts/table.html), that used similar hardware.

* Is HyperApp slow? Not at all. It's faster than some, and slower than others, but the difference is what matters, and that difference is minimal.

* Is HyperApp heavy? HyperApp shines here. Memory usage after page load and after benchmarks remained consistent, even beats Inferno.

* What's the next action? Improve HyperApp's vdom engine. Focus on the first detected weakness: deleting items.

**Comment by krausest** [comment]
To me the most interesting result is create 1,000 rows where hyperapp is twice as fast as all the others whereas create 10,000 is in the range of others. I think you managed to trick my benchmark runner in this case I'm afraid I wouldn't take that result as granted...

**Comment by jbucaran** [comment]
@krausest Did you mean just as a fast?

**Comment by jbucaran** [comment]
## Update 

Removing items is fixed and went from 1.83 down to 1.40, 🎉.

I just need to cut a new release now.

<img width="1249" alt="screen shot 2017-02-27 at 12 55 53" src="https://cloud.githubusercontent.com/assets/56996/23348317/33fefad8-fcec-11e6-9e47-c079ccf28699.png">

**Comment by jbucaran** [comment]
Yep, that was my mistake when adding the frame, it's fixed already. 🙇

**Comment by jbucaran** [comment]
Some improvements: https://github.com/hyperapp/hyperapp/pull/124.

![](https://cloud.githubusercontent.com/assets/56996/23520286/9bc0547c-ffbd-11e6-8db8-a80760dc82a1.png)

**Comment by nichoth** [comment]
Just curious, what led you to implement another view renderer instead of using an existing one? File size?

**Comment by jbucaran** [comment]
@nichoth 🤔 I can think of two reasons: one is file size; I am willing to trade some performance for less bytes. If you look at the git history you'll see we went from virtual-dom, to morphodom, then Snabbdom and finally the current custom engine. 

After https://github.com/hyperapp/hyperapp/pull/124 lands we'll be 1869 bytes, <=1500 without the Router. So, 1kb is more of a stretch now, but it's still in the same range.

Also, speed will have improved significantly. See the js-framework-benchmarks above 🔥.

The other reason is: I wanted the entire source of truth in a single place.

Whenever I try to understand how some library X works, I'm often disappointed that I have to go through dozens of modules, also with deps themselves, and sometimes not written by the same person, which usually translates to quite different coding styles.

**Comment by selfup** [comment]
Should we consider using switch statements for perf improvement? They create lookup tables and are optimized for string evaluation especially

**Comment by jbucaran** [comment]
@selfup 🤔  If it makes things any faster, then why not, I guess. Won't switch statements take more space than if else?

**Comment by dodekeract** [comment]
@jbucaran they have low entropy. With gzip it shouldn't matter.

**Comment by jbucaran** [comment]
I think we can close here and continue to track future optimization efforts in #183. 

Since 0.8.0, and with the introduction of keys, our patch/diff algorithm became more complex (but also more powerful and useful). Performance is now around 1.34~1.37, which is still fine.

**Comment by jorgebucaran** [comment]
V2's new patch algorithm (based on [Superfine](https://github.com/jorgebucaran/superfine)) is significantly faster than V1. Here are the benchmark results. To maximize performance, the benchmarked example app makes use of memoization to boost the rendering perf of table rows that are not affected by state changes. This technique will eventually make it into core as described in #721.

![Superfine](https://user-images.githubusercontent.com/56996/42416263-ee50cec2-82a4-11e8-99fd-021a23e34f05.jpg)
